src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -0,0 +1,171 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "tcp_sockcm_ep.h"
+#include <ucs/sys/sock.h>
+#include <ucs/async/async.h>
+#include <ucs/arch/bitops.h>
+#include <ucs/sys/string.h>
+
+
+static UCS_F_ALWAYS_INLINE
+uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
+{
+    /* return the tcp sockcm connection manager this ep is using */
+    return ucs_container_of(cep->super.super.iface, uct_tcp_sockcm_t, super.iface);
+}
+
+ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_server_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(cep);
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr *server_addr;
+    ucs_async_context_t *async_ctx;
+    uct_tcp_sa_arg_t *sa_arg_ctx;
+    ucs_status_t status;
+    int events;
+
+    cep->state                   |= UCT_TCP_SOCKCM_EP_ON_CLIENT |
+                                    UCT_TCP_SOCKCM_EP_INIT;
+    cep->wireup.client.connect_cb = params->sockaddr_connect_cb.client;
+
+    server_addr = params->sockaddr->addr;
+    status = ucs_socket_create(server_addr->sa_family, SOCK_STREAM, &cep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the fd to non-blocking mode. (so that connect() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(cep->fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    /* try to connect to the server */
+    status = ucs_socket_connect(cep->fd, server_addr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        goto err_close_socket;
+    }
+    ucs_assert ((status == UCS_OK) || (status == UCS_INPROGRESS));
+
+    if (status == UCS_OK) {
+        cep->state &= ~UCT_TCP_SOCKCM_EP_INIT;
+        cep->state |= UCT_TCP_SOCKCM_EP_CONNECTED;
+        /* TODO: start sending the user's data */
```

Comment:
1) can we just call `uct_tcp_sa_data_handler` or its part broken out to separate function to avoid code duplication?
2) pls rename `UCT_TCP_SOCKCM_EP_INIT` to `UCT_TCP_SOCKCM_EP_<CONNECTING|STARTED>` or similar and don't reset it as we do in rdmacm

Reply:
1. what do you mean? why call the async handler here?
2. INIT is before connect() was even called.
we dont have to do the same things we do in rdmacm. why is dropping this specific flag necessary? keeping it would complicate the switch-cases in the async handler a bit.

Reply:
1. because it does (almost) the same things, just to avoid code duplication
2. 
> we dont have to do the same things we do in rdmacm

uct_ep_disconnect behavior is defined in documentation, maybe different way but behavior should be the same

> INIT is before connect() was even called.

then why it's needed? Is this client specific? 
> keeping it would complicate the switch-cases in the async handler a bit.

Is not following doing the same as you have now?
```
swich (state)  {
case UCT_TCP_SOCKCM_EP_ON_CLIENT:
...
case UCT_TCP_SOCKCM_EP_ON_CLIENT | UCT_TCP_SOCKCM_EP_CONNECTED:
...
}
```

Reply:
1. it seems that they are the same for now, but uct_tcp_sa_data_handler() will not do what it does now, in the first case, after connect() completes, it will send data - it wont close everything like it does now.

2. keeping all the flags isn't part of the API...
also, keeping all the flags, will make the 'case' lines very long.
if when starting send i'll set the SENDING flag and once finished sending, will drop it and reset to DATA_SENT, how would it affect ep_disconnect?

Reply:
1. Ok, just move pls common code to new function
2. > how would it affect ep_disconnect?

need to implement behavior documented in https://github.com/openucx/ucx/pull/4586
I'm guessing you are going to do it similar way like we had in rdmacm (based on EP state), so you might face with the same issues like we had recently in rdmacm before we removed all flag drops.

I dont see the problem to have `SENDING | DATA_SENT` flags in the state. it's clear for me that data has been sent (ep started and completed data sending successfully)

If you care about "switch/case" statements, just add combined values which you are expecting to handle to enum:
```
enum {
    flag1 = UCS_BIT(0),
    flag2 = UCS_BIT(1),
    state1 = flag1 | flag2,
    ...
}
```